samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/queue/delivery/TimedAssessmentQueue.java

```diff
@@ -19,92 +19,101 @@
 import java.util.concurrent.ConcurrentHashMap;
-
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.ResourceBundle;
+import org.sakaiproject.component.cover.ServerConfigurationService;
+import org.sakaiproject.event.cover.EventTrackingService;
+import org.sakaiproject.event.cover.NotificationService;
+import org.sakaiproject.samigo.util.SamigoConstants;
+import org.sakaiproject.tool.api.Session;
+import org.sakaiproject.tool.assessment.ui.model.delivery.TimedAssessmentGradingModel;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
-import org.sakaiproject.tool.assessment.ui.model.delivery.TimedAssessmentGradingModel;
+
 
 /**
- * <p>Title: </p>
- * <p>Description: </p>
- * <p>Copyright: Copyright (c) 2004</p>
- * <p>Company: </p>
- * @author not attributable
- * @version 1.0
+ * <p>Title: TimedAssessmentQueue</p>
+ * <p>Description: A queue of assessment timers executed in a pool of threads</p>
  */
 
 public class TimedAssessmentQueue { 
-  private static TimedAssessmentQueue instance; 
-  private ConcurrentHashMap queue;
-    // private SubmitTimedAssessmentThread thread;
-  private Timer timer;
 
+  private ConcurrentHashMap queue;
+  private HashMap<Long, ScheduledFuture<?>> tasks;
+  private ScheduledThreadPoolExecutor threadPool;
   private static Logger log = LoggerFactory.getLogger(TimedAssessmentQueue.class);
+
+  private static class Loader {
+    static final TimedAssessmentQueue INSTANCE = new TimedAssessmentQueue();
+  }
+
   private TimedAssessmentQueue() { 
-    queue = new ConcurrentHashMap ();
+    queue = new ConcurrentHashMap();
+    tasks = new HashMap();
+    // Get any custom thread count, or default to 4
+    int threads = ServerConfigurationService.getInt("samigo.timerThreadCount", 4);
+    log.info( "SAMIGO_TIMED_ASSESSMENT:QUEUE:INIT: THREADS:" + threads + " (Set property 'samigo.timerThreadCount' to adjust.)");
+    threadPool = new ScheduledThreadPoolExecutor(threads);
+    // Set the default removal policy so we don't leak memory with finished tasks
+    threadPool.setRemoveOnCancelPolicy(true);
   } 
 
-  public static synchronized TimedAssessmentQueue getInstance() { 
-    if (instance==null) { 
-      instance = new TimedAssessmentQueue(); 
-    } 
-    return instance; 
+
+  // Get the instance of this class
+  public static TimedAssessmentQueue getInstance() { 
+    return Loader.INSTANCE;
   } 
+
   
+  // Add a timed assessment to the queue
   public void add(TimedAssessmentGradingModel timedAG){
-	  synchronized (queue) {
-		  queue.put(timedAG.getAssessmentGradingId(), timedAG);
-		  log.debug("***1. TimedAssessmentQueue.add, before schedule timer="+timer);
-		  scheduleTask();
-		  log.debug("***2. TimedAssessmentQueue.add, after schedule timer="+timer);
-	  }
+          // Add grading data to the queue
+          synchronized (queue) {
```

Comment:
synchronized on a ConcurentHashMap.put() is not needed as this method is atomic.